text/0018-message-ordering-guarantee.md

```diff
@@ -0,0 +1,171 @@
+}
+```
+
+In this example, without strict ordering, the `ChangeOwnership()` call may happen first, followed by `Spawn()`, followed thereafter by the two RPCs. Such an ordering would result in errors from the `ChangeOwnership()` call (due to the object not yet existing) and the second RPC (due to the ownership not having been changed). Ensuring that all messages arrive in the same order they were sent also ensures that they are called in the order required to process these four operations successfully.
+
+Another example:
+
+```C#
+void FixedUpdate()
+{
+    SendFixedUpdateClientRpc();
+}
+
+void Update()
+{
+    SendUpdateClientRpc();
+}
+```
+
+In this example, the client will execute `SendFixedUpdateClientRpc()` during the `FixedUpdate` stage, and `SendUpdateClientRpc()` during the `Update` stage - this ensures behavior between the sender and the recipient is somewhat more predictable, as logic that occurs on one side during `FixedUpdate` will implicitly occur on the other side during the same stage unless requested otherwise. 
+
+Likewise, the same is true of internal commands:
+
+```C#
+GameObject m_obj;
+
+void Update()
+{
+    m_obj = GameObject.Instantiate(prefab, location, rotation);
+    NetworkBehavior networkBehavior = m_obj.GetComponent<NetworkBehavior>();
+    networkBehavior.Spawn();
+}
+
+void LateUpdate()
+{
+    ulong newOwnerClientId = GetNewOwnerClientId();
+    NetworkBehavior networkBehavior = m_obj.GetComponent<NetworkBehavior>();
+    networkBehavior.ChangeOwnership(newOwnerClientId);
+}
+```
+
+In this example, because `Spawn()` is called during `Update()` on the server, and `ChangeOwnership()` is called during `LateUpdate()` on the server, the same will be true on the client - the object will be spawned on the client during the `Update` stage, and its ownership changed during `LateUpdate`.
+
+# Reference-level explanation
+
+There are two main refactors involved in implementing this change:
+
+1. The first and largest refactor is a change in the way internal messages are sent to ensure they use the same mechanism as RPCs, and the corresponding refactor of that mechanism to ensure it's sufficiently generic to support all message types.
+
+2. The second, smaller refactor is changing the default value of the UpdateStage value in the send parameters so that it reads from the current update stage rather than defaulting to 0 (Update).
+
+The vast majority of the work is in the former refactor, but the biggest user-facing change is the latter one.
+
+In order to complete this refactor, a few things need to happen:
+
+- All RPC-related classes will need to be renamed from Rpc\* to Message\*. This accounts for RpcFrameQueueItem, RpcQueueContainer, RpcQueueProcessor, RpcBatcher, and RpcQueueHistoryFrame.
+
+- All internal commands must be sent through `RpcQueueProcessor.SendFrameQueueItem()`. As a consequence, this also means all internal commands must now build `RpcFrameQueueItem` structs.
```

Comment:
This might be a good place to add an "RpcFrameQueueItem" (perhaps call this MessageFrameQueueItem) pool to avoid memory allocations.  Similar to NetworkBuffers, one could just "get" an already existing struct from the pool.  This would need the ability to "grow the pool" on demand as well as the ability to specify how large you want the pool to be.  (just a thought on this)

Reply:
I think this is a great idea!

Reply:
Per our conversation - opted not to do any sort of caching as a part of this RFC as this RFC is primarily concerned with the sender side of things, and the primary benefit of caching would be on the receiver side. However, the sender is creating things it's not using, so to reduce allocations while I'm working on the sender side, I can create a new struct for send use that doesn't come with unused allocations. I've added that to the RFC.